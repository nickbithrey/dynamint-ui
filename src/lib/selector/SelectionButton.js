export default class SelectionButton {
	
	constructor(text, path, condition, stateFn) {
		this.text = text;
		this.path = path;
		this.condition = condition;
		this.stateFn = stateFn;
	}
	
	build() {
		return {
			text: this.text,
			pathname: this.path,
			stateFn: this.stateFn ? this.stateFn : () => {},
			condition: this.condition ? this.condition : alwaysShow
		};
	}
	
}

const alwaysShow = () => true;

const hasSelection = selection => selection.getSelectedCount() > 0;

const getUriFromSelection = selectionDetails => (
	{
		id: selectionDetails.getSelection()[0].reference,
		uri: selectionDetails.getSelection()[0].uri
	}
);

export {
	alwaysShow,
	hasSelection,
	getUriFromSelection
};