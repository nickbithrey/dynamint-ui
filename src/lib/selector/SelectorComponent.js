import React from 'react';
import PropTypes from 'prop-types';
import { Columnar, Rowise, Block } from '../Grid';
import { Selection } from 'office-ui-fabric-react/lib/DetailsList';
import SelectorTitle from './SelectorTitle';
import SelectorTable from './SelectorTable';
import SelectorButtons from './SelectorButtons';
import SelectionButton, { hasSelection, alwaysShow, getUriFromSelection } from './SelectionButton';

export default class SelectionComponent extends React.Component {
	
	constructor(props) {
		super(props);
		this.onSelectionChanged = this.onSelectionChanged.bind(this);

		this.selection = new Selection({
		    onSelectionChanged: this.onSelectionChanged
	    });
		if (props.includeDefaultButtons) {
			this.buttons = [...props.buttons, ...this.defaultButtons(props.location, props.uri)];
		} else {
			this.buttons = props.buttons;
		}
	}
	
	defaultButtons(location, uri) {
		return [
			new SelectionButton('Edit', location.pathname + '/edit', hasSelection, getUriFromSelection), 
			new SelectionButton('Create', location.pathname + '/create', alwaysShow, deets => ({ isCreate: true, uri: uri }))
		]
	}
	
	onSelectionChanged() {
		this.setState({
			selected: this.selection.getSelection()
		});
	}
	
	render() {
		const {
			title, 
			items, 
			load, 
			columns
		} = this.props;
		return (
			<div>
				<SelectorTitle title={title} />
				<Rowise tokens={{childrenGap: 10}}>
					<Block>
						<SelectorTable items={items} load={load} columns={columns} selection={this.selection} />
					</Block>
					<Block>
						<Columnar horizontalAlign={'end'} tokens={{childrenGap: 5}}>
							<SelectorButtons buttons={this.buttons} selection={this.selection} />
						</Columnar>
					</Block>
				</Rowise>
			</div>
		);
	}
	
}

SelectionComponent.defaultProps = {
	includeDefaultButtons: true,
	buttons: []
}

SelectionComponent.propTypes = {
	title: PropTypes.string,
	items: PropTypes.array,
	load: PropTypes.func,
	columns: PropTypes.arrayOf(PropTypes.object).isRequired,
	buttons: PropTypes.arrayOf(PropTypes.object).isRequired,
	includeDefaultButtons: PropTypes.bool.isRequired
};
