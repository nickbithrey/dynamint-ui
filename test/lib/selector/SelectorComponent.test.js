import React from 'react';
import { shallow } from 'enzyme';

import SelectorComponent from '~/lib/selector/SelectorComponent';
import SelectionButton from '~/lib/selector/SelectionButton';

describe('<SelectionComponent /> rendering', () => {

	const props = {
		title: 'titleText',
		items: [],
		load: jest.fn(() => []),
		columns: [],
		location: {
			pathname: 'path'
		}
	};
	
	let wrapper;
	
	beforeEach(() => {
		wrapper = shallow( <SelectorComponent {...props} /> );
	});

	it('can render standard component with records', () => {
		expect(wrapper).toMatchSnapshot();
	});
		
	it('only renders buttons that have true conditions', () => {
		const newProps = {
			...props,
			buttons: [
				new SelectionButton('Edit', 'path', jest.fn(sel => false), jest.fn(() => {return {}})), 
				new SelectionButton('Create', 'path', jest.fn(sel => true), jest.fn(() => {return {}}))
			],
			includeDefaultButtons: false
		};
		const wrapper = shallow( <SelectorComponent {...newProps} /> );

		const btn = wrapper.find('SelectorButtons');
		expect(btn.length).toBe(1);
		expect(btn.prop('buttons')).toHaveLength(2);
		expect(btn.prop('buttons')[0]).toBe(newProps.buttons[0]);
		expect(btn.prop('buttons')[1]).toBe(newProps.buttons[1]);
	});
	
	it('renders nothing for when no title is supplied', () => {
		let title = wrapper.find('SelectorTitle');
		expect(title).toHaveLength(1);
		expect(title.prop('title')).toBe(props.title);
		
		wrapper.setProps({title: null});
		
		title = wrapper.find('SelectorTitle');
		expect(title).toHaveLength(1);
		expect(title.prop('title')).toBeNull();
	});
		
	it('renders table with details', () => {
		const newProps = {
			...props,
			items: [
				{text: 'val1', num: 1},
				{text: 'val2', num: 2}
			],
			columns: [
				{fieldName: 'text', name: 'Text'},
				{fieldName: 'num', name: 'Number'}
			]
		};
		const wrapper = shallow( <SelectorComponent {...newProps} /> );

		let table = wrapper.find('SelectorTable');
		expect(table.length).toBe(1);
		expect(table.prop('items')).toBe(newProps.items);
		expect(table.prop('load')).toBe(newProps.load);
		expect(table.prop('columns')).toBe(newProps.columns);
	});
	
	it('re-runs button condition when selecting record in table', () => {
		const newProps = {
			...props,
			items: [
				{text: 'val1', num: 1},
				{text: 'val2', num: 2}
			],
			columns: [
				{fieldName: 'text', name: 'Text'},
				{fieldName: 'num', name: 'Number'}
			],
			buttons: [
				new SelectionButton('Edit', 'path', jest.fn(sel => sel.getSelectedCount() > 0), jest.fn(() => {return {}})), 
				new SelectionButton('Create', 'path', jest.fn(sel => true), jest.fn(() => {return {}}))
			],
			includeDefaultButtons: false
		};
		const wrapper = shallow( <SelectorComponent {...newProps} /> );

		let btn = wrapper.find('SelectorButtons');
		expect(btn.prop('selection').getSelectedCount()).toBe(0);

		// call selection changed event
		const selection = btn.prop('selection');
		selection._items = newProps.items;
		selection.setIndexSelected(1, true, true);
		selection._onSelectionChanged();

		btn = wrapper.find('SelectorButtons');
		expect(btn.prop('selection').getSelectedCount()).toBe(1);
	});
	
});