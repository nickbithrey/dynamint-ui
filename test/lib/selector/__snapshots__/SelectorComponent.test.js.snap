// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectionComponent /> rendering can render standard component with records 1`] = `
<div>
  <SelectorTitle
    title="titleText"
  />
  <Rowise
    tokens={
      Object {
        "childrenGap": 10,
      }
    }
  >
    <Block>
      <SelectorTable
        columns={Array []}
        items={Array []}
        load={[MockFunction]}
        selection={
          Selection {
            "_anchoredIndex": 0,
            "_canSelectItem": [Function],
            "_changeEventSuppressionCount": 0,
            "_exemptedCount": 0,
            "_exemptedIndices": Object {},
            "_getKey": [Function],
            "_isModal": false,
            "_items": Array [],
            "_keyToIndexMap": Object {},
            "_onSelectionChanged": [Function],
            "_selectedItems": null,
            "_unselectableCount": 0,
            "_unselectableIndices": Object {},
            "count": 0,
            "mode": 2,
          }
        }
      />
    </Block>
    <Block>
      <Columnar
        horizontalAlign="end"
        tokens={
          Object {
            "childrenGap": 5,
          }
        }
      >
        <SelectorButtons
          buttons={
            Array [
              SelectionButton {
                "condition": [Function],
                "path": "path/edit",
                "stateFn": [Function],
                "text": "Edit",
              },
              SelectionButton {
                "condition": [Function],
                "path": "path/create",
                "stateFn": [Function],
                "text": "Create",
              },
            ]
          }
          selection={
            Selection {
              "_anchoredIndex": 0,
              "_canSelectItem": [Function],
              "_changeEventSuppressionCount": 0,
              "_exemptedCount": 0,
              "_exemptedIndices": Object {},
              "_getKey": [Function],
              "_isModal": false,
              "_items": Array [],
              "_keyToIndexMap": Object {},
              "_onSelectionChanged": [Function],
              "_selectedItems": null,
              "_unselectableCount": 0,
              "_unselectableIndices": Object {},
              "count": 0,
              "mode": 2,
            }
          }
        />
      </Columnar>
    </Block>
  </Rowise>
</div>
`;
